RFCXXX-Native-Json-Adapters.md

```diff
@@ -0,0 +1,275 @@
+---
+RFC: RFCXXX-Native-Json-Adapters.md
+Author: Jim Truher
+Status: Draft
+SupercededBy: 
+Version: 0.9
+Area: engine
+Comments Due: May 1, 2023
+Plan to implement: Yes
+---
+
+# Improving the User Experience of Native Applications by suggesting improved pipelines
+
+## Motivation
+
+As a user I can interact with native applications and have a more PowerShell-like experience.
+
+### Native Applications which Emit Structured Data
+
+Native command execution usually results in textual output which is a fairly poor experience for our users.
+Some of our users use tools like `Crescendo` to enable those tools to "play nice" in the PowerShell arena,
+but the problem of output handling is still troublesome.
+
+Some applications (docker, kubectl, others) can emit `JSON` which is much easier for our users to convert to useful results in PowerShell,
+but the user still has to know the incantation to have the tool emit `JSON` and then they must pipe that output to `Convert-FromJson`.
+We can improve the experience for the user if we can help them discover these tools and additional parameters.
+If the user types:
+
+```powershell
+PS> docker image ls
+```
+
+we can suggest:
+
+```powershell
+PS> docker image ls --format='{{json .}}' | ConvertFrom-Json | Format-Table
+```
+
+There are two aspects of making this work well:
+
+- Knowing that the the parameter `--format='{{json .}}` is available for the `docker` command.
+- Knowing that the invocation will emit json and then suggest how to convert that to objects.
+
+### Providing feedback to the user about improved pipelines
+
+We can improve the experience by using the `FeedbackProvider` infrastructure and present feedback to the user which provides an improved pipeline.
+
+For example, if the user types:
+
+```powershell
+PS> docker image ls
+```
+
+We can provide the following feedback to the user:
+
+```powershell
+docker image ls --format='{{json .}}' | ConvertFrom-Json
+```
+
+or if the user types the following:
+
+```powershell
+PS> uname -a
+```
+
+and the `jc` utility is available, we can provide feedback with an improved pipeline:
+
+```powershell
+PS> uname -a | jc --uname | ConvertFrom-Json
+```
+
+If the user types:
+
+```powershell
+PS> ipconfig
+```
+
+and if a script which converts the output of `ipconfig` into PowerShell objects named `ipconfig-json` exists, we could suggest the following:
+
+```powershell
+PS> ipconfig | ipconfig-json
+```
+
+This last suggested pipeline could be generalized to include a suggestion for any `<utility>-json` command if found.
+This may encourage the community to create conversion functions/scripts.
```

Comment:
Would these `<utility>-json` be part of PS modules/ are cmdlets? I am just wondering about the customer management of these json adapters